### PR TITLE
First draft at a cucumber guide

### PR DESCRIPTION
**Pull request by dkniffin**

### CODE REVIEW

File: qa/cucumber.md [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+````
+
+Things to note here:
+- A user story should be included at the top. The format should match the project's preference.
+- Any `Given` statements that are common to all scenarios in the feature should be included under `Background`
+- Indent `And` statements below their parent `Given`, `When`, or `Then`
+
+## Step definition files
+
+1. Matchers are be grouped by step type (`Given`, `When`, `Then`)
```
Comment (dacur) [inline qa/cucumber.md]: left out the word "to" here

File: qa/cucumber.md [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+2. Matchers with complex regexes have example comments
+3. The matcher is short
+5. The matcher has at most two valuable parameters
```
Comment (dacur) [inline qa/cucumber.md]: i think you skipped #4

File: qa/cucumber.md [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+When the user clicks "something"
+Then the user sees "something"
+````
+
+### Good
+
+````cucumber
+Given I am logged in
+When I click "something"
+Then the user sees
```
Comment (dacur) [inline qa/cucumber.md]: not sure if you left "something" off here or not at the end of the line

**Comment by dacur** [comment]
nice job, can't wait to see it live in the wiki  :100:

File: qa/cucumber.md (hunk 2 above)
Comment (greg5green) [inline qa/cucumber.md]: By valuable parameters, I assume you mean `|one_param, two_param|` is okay and `|one_param, two_param, three_param|` is not okay. It might be good to rephrase this in a way that is more clear.

File: qa/cucumber.md (hunk 3 above)
Comment (greg5green) [inline qa/cucumber.md]: This should be `Then I see`

File: qa/cucumber.md [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+````cucumber
+Story: Animal Submission
+
+  As a Zoologist
+  I want to add a new animal to the site
+  So that I can share my animal knowledge with the community
+
+  Scenario: successful submission
+  Given I'm on the animal creation page
```
Comment (greg5green) [inline qa/cucumber.md]: The Given, When, Thens should be indented to match the style prescribed above.

You might want to do it for the bad scenario too, but it's less important since it's an example of what not to do

File: qa/cucumber.md [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+      And ...
+
+  Scenario: <scenario 2 name>
+    ...
+````
+
+Things to note here:
+- A user story should be included at the top. The format should match the project's preference.
+- Any `Given` statements that are common to all scenarios in the feature should be included under `Background`
+- Indent `And` statements below their parent `Given`, `When`, or `Then`
```
Comment (greg5green) [inline qa/cucumber.md]: I haven't seen this before, but I think I like it

File: qa/cucumber.md (hunk 1 above)
Comment (greg5green) [inline qa/cucumber.md]: Since given, when, and then don't really matter syntactically once you get past the Gherkin, I think it's more useful to group steps by types of functionality.

For example (bad syntax for brevity's sake, sorry):
```
Given(/^I am at the log in screen$/)
When(/^I enter my credentials$/)
Then(/^I am logged in$/)
Given(/^I am a logged in user$/)
```
I think I'd want `I am a logged in user` to be last, since it's a culmination of two or three of the previously listened steps.

Another contrived example:
```
When(/^I click Log In$/)
Then(/^I have access to our wonderful project$/)
... some other stuff ....
When(/^I click Log Out$/)
Then(/^I don't have access to our wonderful project anymore$/)

vs.

When(/^I click Log In$/)
When(/^I click Log Out$/)
... some other stuff ....
Then(/^I have access to our wonderful project$/)
Then(/^I don't have access to our wonderful project anymore$/)
```
The first way, your bits of functionality are nearby -- I need to do some more stuff with log out? I'm not scrolling around the file looking for other log out stuff, it's all right there in one place.

File: qa/cucumber.md [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+
+## Step definition files
+
+1. Matchers are be grouped by step type (`Given`, `When`, `Then`)
+2. Matchers with complex regexes have example comments
+3. The matcher is short
+5. The matcher has at most two valuable parameters
+6. The parameter variables are clearly named
+7. The body is less than ten lines of code
+8. The body does not call other steps
```
Comment (greg5green) [inline qa/cucumber.md]: You should add a link down to the bigger explanation that is coming on line 214!

File: qa/cucumber.md [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+
+  When I add a new animal
+
+  Then I should see the page for my newly created animal
+  And the notice 'Thank you for your animal submission!'
+````
+
+For more information, see [this blog post](http://benmabey.com/2008/05/19/imperative-vs-declarative-scenarios-in-user-stories.html)
+
+## Avoid `step`, in favor of methods
```
Comment (greg5green) [inline qa/cucumber.md]: I realize you don't have the tips and stuff written for this yet, but what is the intent behind this section?

Is it:
```
# Bad
  When(/^I fill in my username$/) do
    fill_in "username", with: "tester"
  end

  When(/^I fill in my password$/) do
    fill_in "password", with: "very secure password"
  end

  When(/^I enter my authentication information$/) do
    step "I fill in my user name"
    step "I fill in my password"
  end

# Good
  When(/^I fill in my username$/) do
    fill_in "username", with: "tester"
  end

  When(/^I fill in my password$/) do
    fill_in "password", with: "very secure password"
  end

  When(/^I enter my authentication information$/) do
    fill_in "username", with: "tester"
    fill_in "password", with: "very secure password"
  end
```
where it is encouraged to use capybara methods or
```
# Bad
  When(/^I fill in my username$/) do
    fill_in "username", with: "tester"
  end

  When(/^I fill in my password$/) do
    fill_in "password", with: "very secure password"
  end

  When(/^I enter my authentication information$/) do
    step "I fill in my user name"
    step "I fill in my password"
  end

# Good
  # my_helper_module.rb
  def fill_in_username
    fill_in "username", with: "tester"
  end

  def fill_in_password
    fill_in "password", with: "very secure password"
  end

  # my_step_def.rb
  When(/^I fill in my username$/) do
    fill_in_username
  end

  When(/^I fill in my password$/) do
    fill_in_password
  end

  When(/^I enter my authentication information$/) do
    fill_in_username
    fill_in_password
  end
```
where the team has written their own helper methods?

Tbh, I'm not super fond of either suggestion (although I could definitely be convinced one of these was a good idea), but I'd like to know what you were thinking or what resources suggested this before I went on a rant I didn't need to write :D

File: qa/cucumber.md (hunk 1 above)
Comment (greg5green) [inline qa/cucumber.md]: Side note: I bring this up because the only large Cucumber code base I have worked with for a long enough period of time to make an opinion was sorted by Given, When, Then and it was awful to work on. All large Cucumber code bases seem to get tough to work on after a while (which is why Spinach exists), so maybe this other suggested way would fail too and my only other experience was anecdotal so /shrug

Comment (dkniffin) [inline qa/cucumber.md]: I like that suggestion. I'll change the doc to reflect it when I get time.

File: qa/cucumber.md (hunk 2 above)
Comment (dkniffin) [inline qa/cucumber.md]: yes. I copied that from the cucumber doc, but the wording definitely leaves room for improvement.

File: qa/cucumber.md (hunk 4 above)
Comment (dkniffin) [inline qa/cucumber.md]: Yep, thanks. I copy pasted that, but it should definitely follow the other rules.

File: qa/cucumber.md (hunk 5 above)
Comment (dkniffin) [inline qa/cucumber.md]: Yea. I need to double check that this actually works. I saw it somewhere, but haven't actually ran code like that yet.

File: qa/cucumber.md (hunk 7 above)
Comment (dkniffin) [inline qa/cucumber.md]: Essentially, the second option. The blog I pulled that from came from here: http://makandracards.com/makandra/18905-how-to-not-repeat-yourself-in-cucumber-scenarios

In addition to that, I'm thinking we could have some shared SB step definitions. For a while I wasn't sure how you'd accomplish that since (for example) the log in process would be different for each project. I think having a set of methods that get called would be a great way to handle those differences.

File: qa/cucumber.md (hunk 5 above)
Comment (dkniffin) [inline qa/cucumber.md]: This does work, fyi.

File: qa/cucumber.md [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+  Scenario: successful submission
+    Given I'm on the animal creation page
+
+    When I add a new animal
+
+    Then I should see the page for my newly created animal
+      And the notice 'Thank you for your animal submission!'
+````
+
+For more information, see [this blog post](http://benmabey.com/2008/05/19/imperative-vs-declarative-scenarios-in-user-stories.html)
```
Comment (albaer) [inline qa/cucumber.md]: It looks like this PR has stagnated, but after reading that blog post, I don't think choosing declarative over imperative all the time is the right call. On eklatz, Spencer was writing a lot of the cucumber. Moving a lot of the heavy lifting to step definitions means that QA has to write a lot of ruby and use capybara rather than focus on the scenarios. More general and granular step definitions make it easier for someone who isn't even looking at the step definitions to jump in and write tests. I also like the idea of integration tests actually acting like a user, especially thinking about it from a QA perspective. "When I add a new animal" isn't something that can be QAed with really specific steps. Say a month into a project, there's a story where the animal's age should validated to be less than 100. It seems like QA (or anyone) should be able to go in and write a scenario for that situation:
```
  When I fill in 'Name' with 'Alligator'
  And fill in 'Age' with '105'
  And click the Create button
  Then I see 'Age must be less than 100'
```
If we're strictly emphasizing declarative style tests, you can't write a scenario for that situation without writing a specific step that says "When I add a new animal with an age over the maximum" or something like that. Part of the point of this kind of test is that the scenarios don't feel like code and they can be written easily by people who aren't devs. This is extreme, but if we're going to hide all the functionality in steps, why not just use rspec?

Comment (dkniffin) [inline qa/cucumber.md]: Great point. I definitely think the advantage of using cucumber over rspec is that it allows you to write the tests in plain english, and ideally, re-use existing step definitions. That would be nearly impossible with declarative. I don't much care which way we do it, as long as it's consistent.

Can you make an edit to put in what you said above?

Comment (albaer) [inline qa/cucumber.md]: Cool. I'll probably get back to this during labs this week.
